sdk/storage/Azure.Storage.Common/src/Shared/PartitionedUploadExtensions.cs

```diff
@@ -39,84 +38,59 @@ internal static bool TryGetLength(Stream content, out long length)
-                byte[] bytes = arrayPool.Rent(blockSize);
-                try
+                long minRequiredBlockSize = (long)Math.Ceiling((double)stream.Length / Constants.Blob.Block.MaxBlocks);
+                if (blockSize < minRequiredBlockSize)
                 {
-                    int offset = 0;
-                    do
-                    {
-                        // You can ask a stream to read however many bytes, but
-                        // it's only going to read as much as it wants.  We're
-                        // trying to saturate the network so we can live with
-                        // sending more, smaller blocks rather than fewer
-                        // perfectly sized blocks that are bound by local I/O.
-                        if (async)
-                        {
-                            read = await stream.ReadAsync(
-                                bytes,
-                                offset,
-                                blockSize - offset,
-                                cancellationToken)
-                                .ConfigureAwait(false);
-                        }
-                        else
-                        {
-                            cancellationToken.ThrowIfCancellationRequested();
-                            read = stream.Read(bytes, offset, blockSize - offset);
-                        }
-                        offset += read;
-                        absolutePosition += read;
-
-                        // Keep reading until we've got enough to fill a block or
-                        // until we can't read any more
-                    } while (offset < minimumBlockSize && read != 0);
+                    throw new ArgumentException($"Max block size {blockSize} not large enough for Storage to hold {stream.Length}.");
+                }
+                // bring min up to our min required by the service
+                acceptableBlockSize = Math.Max(acceptableBlockSize, minRequiredBlockSize);
+            }
 
-                    // If we read anything, turn it into a StreamPartition and
-                    // return it for staging
-                    if (offset != 0)
-                    {
-                        // The StreamParitition is disposable and it'll be the
-                        // user's responsibility to return the bytes used to our
-                        // ArrayPool
-                        yield return new ChunkedStream(absolutePosition, bytes, offset, arrayPool);
+            do
+            {
+                PooledMemoryStream partition = await PooledMemoryStream.BufferStreamPartitionInternal(
+                    stream,
+                    acceptableBlockSize,
+                    blockSize,
+                    absolutePosition,
+                    arrayPool,
+                    126 * Constants.MB, // abritrary limit on single buffer size in multibuffer stream; granular arrays are less wasteful when stream end is reached
```

Comment:
Should this be defined inside PooledMemoryStream?

Reply:
We can introduce a default, and let our code not specify, but I don't think a value like this should be hard-coded into the stream class.

Reply:
If we don't expect to customize that between blob and datalake then I'd rather introduce default. Less code is better.